File: mux/frame.go
package mux

// Encoding a screen for the wire.
//
// The whole of the frame format is here: how a Cell becomes a colour integer,
// how a row of cells becomes text plus style runs, and how a frame's header is
// built. The MERGE — which rows a subscriber gets and whether the result is a
// keyframe — belongs to the hub's slot (hub/watch.go), because it depends on
// what that one subscriber has not read yet.
//
// Two decisions are worth stating, because both look like premature
// compression and neither is:
//
//   - Text and style are SEPARATE. A row is one string plus a handful of runs,
//     not a cell array. A shell line is 80 cells and one or two styles; an
//     array of per-cell objects would be two orders of magnitude larger for the
//     same information, on every frame, at up to 30 frames a second.
//   - Columns in a run are CELL columns, never indexes into the text. A
//     double-width character is one code point in two cells, so the two
//     coordinate systems diverge at the first CJK glyph, and a client painting
//     runs by code point would tint the wrong half of every line after it. `wd`
//     exists so a client can convert without a width table of its own.

import (
	"encoding/json"

	"github.com/MadAppGang/magmux/protocol"
)

// colorInt encodes one Color as the single integer the wire carries.
//
// The three shapes collapse into one number rather than one object because a
// frame is mostly colour: -1 is the terminal's own default, 0..255 is the
// indexed palette, and anything at or above protocol.ColorTrue is truecolor
// with 0xRRGGBB in its low 24 bits. The high bit is what keeps index 255 and
// the colour #0000FF from being the same number.
func colorInt(c Color) int {
	if c.True {
		return protocol.ColorTrue | int(c.R)<<16 | int(c.G)<<8 | int(c.B)
	}
	if c.Index < 0 {
		return protocol.ColorDefault
	}
	return int(c.Index)
}

// rowStyle is what rowWalk records on its way across a row, for the caller that
// wants more than the text.
//
// It is an accumulator rather than a return value so the ONE cell walk can
// serve both readers: capture asks for text and passes nil, the framer asks for
// text plus style and passes one of these. A second walk would drift on the
// first wide-character or NUL-cell fix, and history, selection and the live
// stream disagreeing about what a row said is a bug nobody would look for.
//
// Reused between rows: reset() keeps the slices and drops their contents, so
// encoding a 24-row screen 15 times a second does not allocate 360 slices a
// second.
type rowStyle struct {
	runs []protocol.Run
	wide []int
	// open says the run at the end of runs is still being extended, and by
	// which style. A default-styled cell closes it: a run is only ever emitted
	// for cells that differ from the terminal's own defaults.
	open         bool
	fg, bg, attr int
	lastCol      int
}

func (st *rowStyle) reset() {
	st.runs = st.runs[:0]
	st.wide = st.wide[:0]
	st.open = false
}

// add folds one cell into the style runs. Merging happens here rather than in a
// second pass because a run is only ever extended by the cell to its right.
func (st *rowStyle) add(col int, c Cell) {
	fg, bg, attr := colorInt(c.Fg), colorInt(c.Bg), int(c.Attr)
	if fg == protocol.ColorDefault && bg == protocol.ColorDefault && attr == 0 {
		st.open = false
		return
	}
	if st.open && st.fg == fg && st.bg == bg && st.attr == attr && st.lastCol+1 == col {
		st.runs[len(st.runs)-1][1]++
		st.lastCol = col
		return
	}
	st.runs = append(st.runs, protocol.NewRun(col, 1, fg, bg, attr))
	st.open, st.fg, st.bg, st.attr, st.lastCol = true, fg, bg, attr, col
}

// encodeLine renders one row of cells as the JSON object a frame carries.
//
// cols bounds the walk, and rowWalk clamps it again to the row itself: a
// scrollback row keeps the width it was printed at and can be wider or narrower
// than the screen showing it, so every reader has to bound by len(row) as well.
func encodeLine(y int, row []Cell, cols int, st *rowStyle) ([]byte, error) {
	st.reset()
	text := rowWalk(row, 0, cols-1, st)
	line := protocol.Line{Y: y, T: text}
	if len(st.wide) > 0 {
		line.Wd = append([]int(nil), st.wide...)
	}
	if len(st.runs) > 0 {
		line.R = append([]protocol.Run(nil), st.runs...)
	}
	return json.Marshal(line)
}

// frameHeader builds the part of a frame that is fixed when the screen is READ.
//
// It hands back a JSON object that has been opened and NOT closed, with no
// trailing comma, because the hub appends the two fields it owns — `key` and
// `lines` — and only it knows what they are once a keyframe and the deltas
// behind it have merged into one message. json.Marshal of a struct always ends
// in '}', so dropping the last byte is exact rather than a parse.
func frameHeader(h protocol.FrameHeader) ([]byte, error) {
	b, err := json.Marshal(h)
	if err != nil {
		return nil, err
	}
	return b[:len(b)-1], nil
}

// changedLine is the notify mode's whole message: this pane moved, go and read
// it. It is a complete line, newline included, because it travels through the
// same slot a frame does and a slot's raw form is written verbatim.
func changedLine(pane int) []byte {
	b, err := json.Marshal(map[string]any{"type": protocol.EventChanged, "pane": pane})
	if err != nil {
		return nil
	}
	return append(b, '\n')
}

File: demo/rc/frame.ts
/**
 * The ONE frame decoder for both demo clients.
 *
 * Pure: no I/O, no DOM, no ANSI, no `node:` import. It is imported unchanged by
 * `cli.ts` (bun) and by `web/app.ts` (bundled for a browser by `serve.ts`), so
 * the two surfaces cannot drift in their reading of the wire. What it does NOT
 * do is decide how a colour or an attribute is PAINTED — `cli.ts` turns the
 * classification below into SGR and `app.ts` turns it into CSS, because that is
 * surface, not protocol.
 *
 * The wire shape is `protocol/frame.go`, and three of its rules are the whole
 * of this file:
 *
 *   1. A frame is WHOLE ROWS. `key:true` means `lines` covers every row and the
 *      client clears first; `key:false` means each line REPLACES the row it
 *      names. So deltas merge by row index and a gap in `seq` is harmless.
 *   2. Run columns are CELL columns, not indexes into `t`. A double-width
 *      glyph is one code point in two cells, and `wd` lists the code-point
 *      indexes that are wide. A client that painted by code point would drift
 *      one column right of every CJK glyph on the line.
 *   3. `t` is right-trimmed and the runs are NOT. A run may extend past the end
 *      of the text — that is how a styled blank (a selection, a filled status
 *      bar) survives the trim.
 *
 * A frame with no `lines` at all is legitimate and must not be discarded: a
 * moved or hidden cursor, a scrollback change and an alt-screen switch all
 * change what a client draws without changing a cell.
 */

// ── wire types (protocol/frame.go) ──────────────────────────────────────────

/** One stretch of identically-styled cells: [col, len, fg, bg, attr]. */
export type Run = [number, number, number, number, number];

export interface Line {
  y: number;
  /** The row's text, one code point per non-continuation cell, right-trimmed. */
  t: string;
  /** Code-point indexes in `t` that occupy TWO cells. */
  wd?: number[];
  r?: Run[];
}

export interface Cursor {
  y: number;
  x: number;
  vis: boolean;
}

export interface FrameMessage {
  type: string;
  pane: number;
  seq: number;
  rows: number;
  cols: number;
  alt: boolean;
  sb: number;
  scrolled: boolean;
  cur: Cursor;
  key: boolean;
  lines?: Line[];
}

// ── colour classification ───────────────────────────────────────────────────

/** No colour: use the surface's own foreground/background. */
export const COLOR_DEFAULT = -1;
/** Anything at or above this is truecolor; the low 24 bits are the RGB. */
export const COLOR_TRUE = 1 << 24;
export const COLOR_RGB_MASK = 0xffffff;

export type Colour =
  | { kind: "default" }
  | { kind: "indexed"; index: number }
  | { kind: "rgb"; r: number; g: number; b: number };

const DEFAULT_COLOUR: Colour = { kind: "default" };

/**
 * Classify one wire colour. CLASSIFICATION ONLY — what a client does with the
 * answer is its own business, and the two painters disagree about it on
 * purpose.
 */
export function classifyColour(c: number): Colour {
  if (c >= COLOR_TRUE) {
    const rgb = c & COLOR_RGB_MASK;
    return { kind: "rgb", r: (rgb >> 16) & 0xff, g: (rgb >> 8) & 0xff, b: rgb & 0xff };
  }
  if (c >= 0 && c <= 255) return { kind: "indexed", index: c };
  return DEFAULT_COLOUR;
}

// ── attributes ──────────────────────────────────────────────────────────────

/**
 * The nine attribute bits, restated.
 *
 * SOURCE OF TRUTH: `mux/cell.go:10-18` (`Attr`, `AttrBold` … `AttrOverline`).
 * `Run[4]` is public wire data whose bits are defined in a package no external
 * client may import, so every client outside magmux restates them — and this
 * copy is the one that goes stale silently. `selftest.ts` check 1 pins the
 * values; if magmux ever renumbers them, that check is what fails.
 */
export const ATTR_BOLD = 1 << 0; // 1
export const ATTR_DIM = 1 << 1; // 2
export const ATTR_ITALIC = 1 << 2; // 4
export const ATTR_BLINK = 1 << 3; // 8
export const ATTR_REVERSE = 1 << 4; // 16
export const ATTR_INVIS = 1 << 5; // 32
export const ATTR_UNDERLINE = 1 << 6; // 64
export const ATTR_STRIKE = 1 << 7; // 128
export const ATTR_OVERLINE = 1 << 8; // 256

/** Every bit, in wire order, for anything that wants to enumerate them. */
export const ATTR_BITS: { name: string; bit: number }[] = [
  { name: "bold", bit: ATTR_BOLD },
  { name: "dim", bit: ATTR_DIM },
  { name: "italic", bit: ATTR_ITALIC },
  { name: "blink", bit: ATTR_BLINK },
  { name: "reverse", bit: ATTR_REVERSE },
  { name: "invisible", bit: ATTR_INVIS },
  { name: "underline", bit: ATTR_UNDERLINE },
  { name: "strike", bit: ATTR_STRIKE },
  { name: "overline", bit: ATTR_OVERLINE },
];

export function hasAttr(attr: number, bit: number): boolean {
  return (attr & bit) !== 0;
}

// ── the cp <-> cell mapping ─────────────────────────────────────────────────

export interface CellCols {
  /** The row's code points, split once so no caller splits it again. */
  cps: string[];
  /**
   * Cell column each code point starts at. Length is `cps.length + 1`; the
   * last entry is the total cell width of the text, which is where the
   * right-trimmed blanks begin.
   */
  colOf: Int32Array;
  /**
   * Code-point index occupying each cell column, or -1 for the CONTINUATION
   * half of a wide glyph and for every column past the text. Length is
   * `colOf[cps.length]`, so a caller must bound its own walk by the pane's
   * column count and treat anything beyond as blank.
   */
  cpAt: Int32Array;
}

/**
 * Build the cp <-> cell mapping for one line, both directions, from `wd`.
 *
 * `wd` is the only width information on the wire, and taking it rather than
 * measuring the glyph is deliberate: magmux already decided how wide each cell
 * is when it drew the screen, and a client re-deriving it from its own width
 * table would disagree with the pane it is mirroring.
 */
export function cellCols(line: Line): CellCols {
  const cps = Array.from(line.t ?? "");
  const wide = new Set(line.wd ?? []);
  const colOf = new Int32Array(cps.length + 1);
  let col = 0;
  for (let i = 0; i < cps.length; i++) {
    colOf[i] = col;
    col += wide.has(i) ? 2 : 1;
  }
  colOf[cps.length] = col;

  const cpAt = new Int32Array(col).fill(-1);
  for (let i = 0; i < cps.length; i++) cpAt[colOf[i]] = i;
  return { cps, colOf, cpAt };
}

// ── the screen model ────────────────────────────────────────────────────────

export interface Screen {
  rows: number;
  cols: number;
  /** One entry per row, index === y. Never sparse. */
  lines: Line[];
  cur: Cursor;
  seq: number;
  alt: boolean;
  sb: number;
  scrolled: boolean;
  pane: number;
  /** True when the last frame applied was a keyframe: the painter must clear. */
  cleared: boolean;
  /** Frames applied since the screen was created. */
  frames: number;
}

function blankLine(y: number): Line {
  return { y, t: "" };
}

export function newScreen(rows: number, cols: number, pane = 0): Screen {
  return {
    rows,
    cols,
    lines: Array.from({ length: Math.max(0, rows) }, (_, y) => blankLine(y)),
    cur: { y: 0, x: 0, vis: true },
    seq: 0,
    alt: false,
    sb: 0,
    scrolled: false,
    pane,
    cleared: false,
    frames: 0,
  };
}

/**
 * Apply one frame and report which rows changed.
 *
 * A keyframe returns EVERY row index and sets `screen.cleared`, because the
 * client must clear before repainting — `key:true` means the client has no
 * memory of what came before, which is the property that makes a resize, an
 * alt-screen switch and a second watcher joining all safe.
 *
 * A delta returns only the rows it carried. It is not an optimisation to return
 * fewer: a row the frame did not name is a row magmux is asserting did not
 * change, and repainting it would be work, not correctness.
 *
 * GEOMETRY COMES FROM EVERY FRAME, not only from the watch reply. A pane
 * reflows whenever the human drags the window — which in a demo is the most
 * likely thing anyone does — and `FrameHeader` carries `rows` and `cols` on
 * every frame for exactly that reason. A resize and an alt-screen switch are
 * therefore treated as keyframes here even if `key` is false: magmux forces one
 * in both cases, and a client that trusted `key` alone would paint row deltas
 * onto a screen of the wrong shape the one time it did not.
 */
export function applyFrame(screen: Screen, frame: FrameMessage): number[] {
  const rows = frame.rows > 0 ? frame.rows : screen.rows;
  const cols = frame.cols > 0 ? frame.cols : screen.cols;
  const resized = rows !== screen.rows || cols !== screen.cols;
  const switched = screen.frames > 0 && (frame.alt === true) !== screen.alt;
  const key = frame.key === true || resized || switched;

  if (resized) {
    const next: Line[] = Array.from({ length: rows }, (_, y) => screen.lines[y] ?? blankLine(y));
    screen.lines = next;
    screen.rows = rows;
    screen.cols = cols;
  }
  if (key) {
    for (let y = 0; y < screen.rows; y++) screen.lines[y] = blankLine(y);
  }

  const changed: number[] = [];
  for (const line of frame.lines ?? []) {
    const y = line.y;
    if (!Number.isInteger(y) || y < 0 || y >= screen.rows) continue; // a row off this screen is not ours to paint
    screen.lines[y] = line;
    if (!key) changed.push(y);
  }
  if (key) for (let y = 0; y < screen.rows; y++) changed.push(y);

  screen.cur = frame.cur ?? screen.cur;
  screen.seq = frame.seq ?? screen.seq;
  screen.alt = frame.alt === true;
  screen.sb = frame.sb ?? 0;
  screen.scrolled = frame.scrolled === true;
  screen.pane = frame.pane ?? screen.pane;
  screen.cleared = key;
  screen.frames++;
  return changed;
}

// ── row decoding, shared by both painters ───────────────────────────────────

/** One stretch of a row with a single style, ready for a painter. */
export interface Span {
  /** Cell column this span starts at. */
  col: number;
  /** Cell columns it covers — NOT `text.length`, which differs on wide glyphs. */
  width: number;
  text: string;
  fg: number;
  bg: number;
  attr: number;
}

/**
 * Expand a row's runs into a per-cell-column style lookup.
 *
 * Runs arrive merged and in order, but they are allowed to run past the text
 * and are clamped to `cols` here rather than at every read site.
 */
function styleColumns(line: Line, cols: number) {
  const fg = new Int32Array(cols).fill(COLOR_DEFAULT);
  const bg = new Int32Array(cols).fill(COLOR_DEFAULT);
  const attr = new Int32Array(cols);
  for (const r of line.r ?? []) {
    const start = Math.max(0, r[0]);
    const end = Math.min(cols, r[0] + r[1]);
    for (let c = start; c < end; c++) {
      fg[c] = r[2];
      bg[c] = r[3];
      attr[c] = r[4];
    }
  }
  return { fg, bg, attr };
}

/**
 * Decode one row into styled spans over `cols` cell columns.
 *
 * This is the shared half of painting, and it is here rather than in the two
 * painters because getting it wrong is invisible: a continuation cell that
 * contributes a character, or a run boundary read in code points, shifts
 * everything to its right by one column and looks like a rendering glitch
 * rather than a decode bug.
 *
 * Every column 0..cols-1 is covered, blanks included, because a frame's rows
 * are right-trimmed and a painter that stopped at the end of the text would
 * leave the previous frame's pixels standing to its right.
 */
export function rowSpans(line: Line, cells: CellCols, cols: number): Span[] {
  const style = styleColumns(line, cols);
  const spans: Span[] = [];
  let cur: Span | null = null;

  for (let c = 0; c < cols; c++) {
    const cp = c < cells.cpAt.length ? cells.cpAt[c] : -1;
    const isCont = c < cells.cpAt.length && cp < 0; // inside the text, so a continuation half
    if (isCont && cur) {
      // The wide glyph before it already claims this column; widen the span
      // rather than emitting anything, or the row gains a character.
      cur.width++;
      continue;
    }
    const fg = style.fg[c];
    const bg = style.bg[c];
    const attr = style.attr[c];
    const ch = cp >= 0 ? cells.cps[cp] : " ";
    const w = cp >= 0 && cells.colOf[cp + 1] - cells.colOf[cp] === 2 ? 2 : 1;
    if (cur && cur.fg === fg && cur.bg === bg && cur.attr === attr) {
      cur.text += ch;
      cur.width += w;
    } else {
      cur = { col: c, width: w, text: ch, fg, bg, attr };
      spans.push(cur);
    }
    if (w === 2) c++; // the continuation half is already accounted for
  }
  return spans;
}

/** The plain text of a row over `cols` columns — what a viewer would read. */
export function rowText(line: Line, cols: number): string {
  const cells = cellCols(line);
  let out = "";
  let c = 0;
  while (c < cols) {
    const cp = c < cells.cpAt.length ? cells.cpAt[c] : -1;
    if (cp >= 0) {
      out += cells.cps[cp];
      c += cells.colOf[cp + 1] - cells.colOf[cp];
      continue;
    }
    if (c < cells.cpAt.length) {
      c++; // a continuation half, already emitted with its glyph
      continue;
    }
    out += " ";
    c++;
  }
  return out;
}

/** Every row of a screen as text, right-trimmed — for tests and diagnostics. */
export function screenText(screen: Screen): string {
  return screen.lines.map((l) => rowText(l, screen.cols).replace(/ +$/, "")).join("\n");
}
